app/models/warehouse/build_report.rb

```diff
@@ -0,0 +1,66 @@
+module Warehouse
+  class BuildReport
+    def run!
+      # we just flush the whole damned thing every time
+      Hmis::Report.delete_all
+      at = MatchDecisions::Base.arel_table
+      ::Client.joins( :project_client, client_opportunity_matches: :decisions ).preload(
+        :project_client,
+        {
+          client_opportunity_matches: {
+            decisions: %i[
+              decline_reason
+              not_working_with_client_reason
+              administrative_cancel_reason
+            ]
+          }
+        }
+      ).where( at[:status].not_eq nil ).distinct.find_each do |client|
+        client_id = client.project_client.id_in_data_source
+        client.client_opportunity_matches.each do |match|
+          # similar to match.current_decision, but more efficient given that we've preloaded all the decisions
+          decisions = match.decisions.select(&:status).sort_by(&:created_at)
```

Comment:
This is not actually redundant. The query ignores clients that have only decisions with null statuses, but the preload then fetches in *all* decisions, so we pare them out here.